Assumption test fails with geth 1.7.3

## Problem Definition

After upgrading to `geth 1.7.3-stable` one of the *assumption* tests fails. See log below. 

```
% pytest raiden/tests/integration/rpc/                                                   17:51:44
======================================= test session starts ========================================
platform darwin -- Python 2.7.14, pytest-3.2.2, py-1.4.34, pluggy-0.4.0
rootdir: /Users/paul/Work/raiden, inifile: tox.ini
plugins: timeout-1.2.0, random-0.2, cov-2.5.1, catchlog-1.2.2, hypothesis-3.33.0, flaky-3.4.0
timeout: 540.0s method: signal
collected 7 items

raiden/tests/integration/rpc/test_assumptions.py ....F..
===Flaky Test Report===


===End Flaky Test Report===

============================================= FAILURES =============================================
________________________________ test_transact_throws_opcode[geth] _________________________________

deploy_client = <JSONRPCClient @29878>
blockchain_backend = [<Popen at 0x107554550 pid=39539 returncode=None>]

    @pytest.mark.parametrize('blockchain_type', ['geth'])
    def test_transact_throws_opcode(deploy_client, blockchain_backend):
        """ The receipt status field of a transaction that threw is 0x0 """
        contract_proxy = deploy_rpc_test_contract(deploy_client)

        address = contract_proxy.address
        assert deploy_client.eth_getCode(address) != '0x'

>       gas = min(contract_proxy.fail.estimate_gas(), deploy_client.gaslimit())

raiden/tests/integration/rpc/test_assumptions.py:99:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
raiden/network/rpc/smartcontract_proxy.py:113: in estimate_gas
    **kargs
raiden/network/rpc/client.py:660: in eth_estimateGas
    res = self.call('eth_estimateGas', json_data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <JSONRPCClient @29878>, method = 'eth_estimateGas'
args = ({'data': '0xa9cc4718', 'from': '0x8d2333af216c296834d968d1af88ff812061eabd', 'gas': '0x4a817c800', 'gasPrice': '0x4a817c800', ...},)
request = <tinyrpc.protocols.jsonrpc.JSONRPCRequest object at 0x107554850>
reply = '{"jsonrpc":"2.0","id":17,"error":{"code":-32000,"message":"gas required exceeds allowance or always failing transaction"}}\n'
jsonrpc_reply = <tinyrpc.protocols.jsonrpc.JSONRPCErrorResponse object at 0x107554750>

    def call(self, method, *args):
        """ Do the request and return the result.

            Args:
                method (str): The RPC method.
                args: The encoded arguments expected by the method.
                    - Object arguments must be supplied as a dictionary.
                    - Quantity arguments must be hex encoded starting with '0x' and
                    without left zeros.
                    - Data arguments must be hex encoded starting with '0x'
            """
        request = self.protocol.create_request(method, args)
        reply = self.transport.send_message(request.serialize())

        jsonrpc_reply = self.protocol.parse_reply(reply)
        if isinstance(jsonrpc_reply, JSONRPCSuccessResponse):
            return jsonrpc_reply.result
        elif isinstance(jsonrpc_reply, JSONRPCErrorResponse):
>           raise EthNodeCommunicationError(jsonrpc_reply.error)
E           EthNodeCommunicationError: gas required exceeds allowance or always failing transaction

raiden/network/rpc/client.py:470: EthNodeCommunicationError
-------------------------------------- Captured stdout setup ---------------------------------------
Your new account is locked with a password. Please give a password. Do not forget this password.
!! Unsupported terminal, password will be echoed.
Passphrase:
Repeat passphrase:
Address: {8d2333af216c296834d968d1af88ff812061eabd}
Unlocking account 0 | Attempt 1/3
!! Unsupported terminal, password will be echoed.
Passphrase:
-------------------------------------- Captured stderr setup ---------------------------------------
WARN [11-21|17:55:16] No etherbase set and no accounts found as default
=============================== 1 failed, 6 passed in 56.62 seconds ================================
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,9 +7,9 @@ python:
 - '3.6'
 env:
   global:
-  - GETH_URL_LINUX='https://gethstore.blob.core.windows.net/builds/geth-linux-amd64-1.7.2-1db4ecdc.tar.gz'
-  - GETH_URL_MACOS='https://gethstore.blob.core.windows.net/builds/geth-darwin-amd64-1.7.2-1db4ecdc.tar.gz'
-  - GETH_VERSION='1.7.2'
+  - GETH_URL_LINUX='https://gethstore.blob.core.windows.net/builds/geth-linux-amd64-1.7.3-4bb3c89d.tar.gz'
+  - GETH_URL_MACOS='https://gethstore.blob.core.windows.net/builds/geth-darwin-amd64-1.7.3-4bb3c89d.tar.gz'
+  - GETH_VERSION='1.7.3'
   - SOLC_URL_LINUX='https://github.com/ethereum/solidity/releases/download/v0.4.18/solc-static-linux'
   - SOLC_URL_MACOS='https://www.dropbox.com/s/21xuco6djjpdwd4/solc_0.4.18?dl=0'
   - SOLC_VERSION='v0.4.18'

--- a/raiden/exceptions.py
+++ b/raiden/exceptions.py
@@ -152,7 +152,9 @@ class RaidenShuttingDown(RaidenError):
 class EthNodeCommunicationError(RaidenError):
     """ Raised when something unexpected has happened during
     communication with the underlying ethereum node"""
-    pass
+    def __init__(self, error_msg, error_code=None):
+        super(EthNodeCommunicationError, self).__init__(error_msg)
+        self.error_code = error_code
 
 
 class AddressWithoutCode(RaidenError):

--- a/raiden/network/rpc/client.py
+++ b/raiden/network/rpc/client.py
@@ -496,7 +496,7 @@ class JSONRPCClient:
         if isinstance(jsonrpc_reply, JSONRPCSuccessResponse):
             return jsonrpc_reply.result
         elif isinstance(jsonrpc_reply, JSONRPCErrorResponse):
-            raise EthNodeCommunicationError(jsonrpc_reply.error)
+            raise EthNodeCommunicationError(jsonrpc_reply.error, jsonrpc_reply._jsonrpc_error_code)
         else:
             raise EthNodeCommunicationError('Unknown type of JSONRPC reply')
 
@@ -689,12 +689,9 @@ class JSONRPCClient:
         try:
             res = self.call('eth_estimateGas', json_data)
         except EthNodeCommunicationError as e:
-            tx_would_fail = str(e) in [
-                'gas required exceeds allowance or always failing transaction',  # geth
-                'Transaction execution error.'  # parity
-            ]
-            if tx_would_fail:
-                return -1
+            tx_would_fail = e.error_code and e.error_code in (-32015, -32000)
+            if tx_would_fail:  # -32015 is parity and -32000 is geth
+                return None
             else:
                 raise e
 

--- a/raiden/network/rpc/client.py
+++ b/raiden/network/rpc/client.py
@@ -686,7 +686,17 @@ class JSONRPCClient:
             startgas,
             gasprice,
         )
-        res = self.call('eth_estimateGas', json_data)
+        try:
+            res = self.call('eth_estimateGas', json_data)
+        except EthNodeCommunicationError as e:
+            tx_would_fail = str(e) in [
+                'gas required exceeds allowance or always failing transaction',  # geth
+                'Transaction execution error.'  # parity
+            ]
+            if tx_would_fail:
+                return -1
+            else:
+                raise e
 
         return quantity_decoder(res)
 

--- a/raiden/network/rpc/smartcontract_proxy.py
+++ b/raiden/network/rpc/smartcontract_proxy.py
@@ -77,6 +77,12 @@ class ContractProxy:
         return res
 
     def estimate_gas(self, function_name: str, *args, **kargs):
+        """ Returns the estimated gas for the function or -1 if the function
+        will throw.
+
+        Raises:
+            EthNodeCommunicationError: If the ethereum node's reply can't be parsed.
+        """
         if not self.estimate_function:
             raise RuntimeError('estimate_function was not supplied.')
 

--- a/raiden/network/rpc/transactions.py
+++ b/raiden/network/rpc/transactions.py
@@ -32,7 +32,7 @@ def estimate_and_transact(proxy, function_name, startgas, gasprice, *args):
 
     # XXX: From Byzantium and on estimate gas is not giving an accurate estimation
     #      and as such we not longer utilize its result but use the GAS_LIMIT in
-    #      all transactions. With the revert() call not consumin all given gas that
+    #      all transactions. With the revert() call not consuming all given gas that
     #      is not that bad
     #
     # estimated_gas = callobj.estimate_gas(

--- a/raiden/tests/integration/rpc/test_assumptions.py
+++ b/raiden/tests/integration/rpc/test_assumptions.py
@@ -84,7 +84,7 @@ def test_estimate_gas_fail(deploy_client, blockchain_backend):
     address = contract_proxy.contract_address
     assert len(deploy_client.eth_getCode(address)) > 0
 
-    assert contract_proxy.estimate_gas('fail') == -1
+    assert not contract_proxy.estimate_gas('fail')
 
 
 def test_transact_opcode(deploy_client, blockchain_backend):

--- a/raiden/tests/integration/rpc/test_assumptions.py
+++ b/raiden/tests/integration/rpc/test_assumptions.py
@@ -77,6 +77,16 @@ def test_call_throws(deploy_client, blockchain_backend):
     assert contract_proxy.call('fail') == b''
 
 
+def test_estimate_gas_fail(deploy_client, blockchain_backend):
+    """ A JSON RPC estimate gas call for a throwing transaction returns -1"""
+    contract_proxy = deploy_rpc_test_contract(deploy_client)
+
+    address = contract_proxy.contract_address
+    assert len(deploy_client.eth_getCode(address)) > 0
+
+    assert contract_proxy.estimate_gas('fail') == -1
+
+
 def test_transact_opcode(deploy_client, blockchain_backend):
     """ The receipt status field of a transaction that did not throw is 0x1 """
     contract_proxy = deploy_rpc_test_contract(deploy_client)

--- a/raiden/tests/integration/rpc/test_assumptions.py
+++ b/raiden/tests/integration/rpc/test_assumptions.py
@@ -101,10 +101,7 @@ def test_transact_throws_opcode(deploy_client, blockchain_backend):
     address = contract_proxy.contract_address
     assert len(deploy_client.eth_getCode(address)) > 0
 
-    gas = min(
-        contract_proxy.estimate_gas('fail'),
-        deploy_client.gaslimit(),
-    )
+    gas = deploy_client.gaslimit()
 
     transaction_hex = contract_proxy.transact('fail', startgas=gas)
     transaction = unhexlify(transaction_hex)

--- a/tools/testnet/files/dockerfiles/geth-testnet/Dockerfile
+++ b/tools/testnet/files/dockerfiles/geth-testnet/Dockerfile
@@ -1,4 +1,4 @@
-FROM ethereum/client-go:v1.7.2
+FROM ethereum/client-go:v1.7.3
 MAINTAINER Ulrich Petri <ulrich@brainbot.com>
 
 RUN \

